File: tests/lib/collections/collectionBuilder.ts
import { type Style, STYLES, Video } from '../../../src/lib/collections/video';
import { fakerFR } from '@faker-js/faker';
import { Collection } from '../../../src/lib/collections/collection.ts';

interface Builder<T> {
  build: () => T;
}

class CollectionBuilder implements Builder<Collection> {
  private id: string = fakerFR.string.uuid();
  private title: string = fakerFR.lorem.sentence();
  private videos: Video[] = [];
  build(): Collection {
    return { id: this.id, title: this.title, videos: this.videos };
  }

  withId(id: string) {
    this.id = id;
    return this;
  }

  withTitle(title: string) {
    this.title = title;
    return this;
  }

  addVideo(video: Video) {
    this.videos.push(video);
    return this;
  }
}

class VideoBuilder implements Builder<Video> {
  private path: string = fakerFR.system.filePath();
  private thumbnail: string = fakerFR.string.alpha();
  private size: number = fakerFR.number.int();
  private name: string = fakerFR.string.alpha();
  private artist: string = fakerFR.music.artist();
  private song: string = fakerFR.music.songName();
  private style: Style[] = [];
  private tags: string[] = [];

  constructor() {
    const range = fakerFR.number.int({ min: 0, max: STYLES.length - 1 });
    for (let i = 0; i < range; i++) {
      this.style.push(
        STYLES[fakerFR.number.int({ min: 0, max: STYLES.length - 1 })]
      );
      this.tags.push(fakerFR.lorem.word());
    }
  }

  withName(name: string) {
    this.name = name;
    return this;
  }

  withArtist(artist: string) {
    this.artist = artist;
    return this;
  }

  withSong(song: string) {
    this.song = song;
    return this;
  }

  withTags(tags: string[]) {
    this.tags = tags;
    return this;
  }

  withStyles(styles: Style[]) {
    this.style = styles;
    return this;
  }

  withPath(path: string) {
    this.path = path;
    return this;
  }

  withSize(size: number) {
    this.size = size;
    return this;
  }

  withThumbnail(thumbnail: string) {
    this.thumbnail = thumbnail;
    return this;
  }

  build(): Video {
    return {
      play(): string {
        return '';
      },
      toHumanReadable(): string {
        return '';
      },
      path: this.path,
      thumbnail: this.thumbnail,
      size: this.size,
      name: this.name,
      artist: this.artist,
      song: this.song,
      style: this.style,
      tags: this.tags,
    };
  }
}
export const aVideo = () => new VideoBuilder();

export const aCollection = () => new CollectionBuilder();

File: src-tauri/src/collections/collection_builder.rs
#[cfg(test)]
use crate::collections::collections::{Collection, Style, Video};
#[cfg(test)]
use fake::faker::filesystem::fr_fr::FilePath;
#[cfg(test)]
use fake::faker::job::fr_fr::Title;
#[cfg(test)]
use fake::faker::name::fr_fr::Name;
#[cfg(test)]
use fake::Fake;
#[cfg(test)]
use fake::Faker;
#[cfg(test)]
use std::path::PathBuf;

#[cfg(test)]
pub struct CollectionBuilder {
    title: String,
    videos: Vec<Video>,
}

#[cfg(test)]
impl CollectionBuilder {
    fn new() -> CollectionBuilder {
        CollectionBuilder {
            title: "".to_string(),
            videos: vec![],
        }
    }

    pub fn with_title(mut self, title: &str) -> Self {
        self.title = title.parse().unwrap();
        self
    }

    pub fn add_video(mut self, video: &Video) -> Self {
        self.videos.push(video.clone());
        self
    }

    pub fn build(self) -> Collection {
        Collection {
            id: uuid::Uuid::new_v4(),
            title: self.title.clone(),
            videos: self.videos.clone(),
        }
    }
}

#[cfg(test)]
pub struct VideoBuilder {
    path: String,
    name: String,
    artist: String,
    song: String,
    style: Vec<Style>,
    tags: Vec<String>,
    thumbnail: String,
    size_bytes: u64,
}

#[cfg(test)]
impl VideoBuilder {
    fn new() -> VideoBuilder {
        VideoBuilder {
            path: FilePath().fake(),
            name: Title().fake(),
            artist: Name().fake(),
            song: Title().fake(),
            style: vec![],
            tags: vec![],
            thumbnail: FilePath().fake(),
            size_bytes: Faker.fake::<u64>(),
        }
    }

    pub fn build(self) -> Video {
        Video {
            path: PathBuf::from(self.path),
            name: self.name.clone(),
            artist: self.artist.clone(),
            song: self.song.clone(),
            style: self.style.clone(),
            tags: self.tags.clone(),
            thumbnail: self.thumbnail.clone(),
            size_bytes: self.size_bytes,
        }
    }
}

#[cfg(test)]
pub fn a_collection() -> CollectionBuilder {
    CollectionBuilder::new()
}

#[cfg(test)]
pub fn a_video() -> VideoBuilder {
    VideoBuilder::new()
}
